File: app/src/main/java/com/battlelancer/seriesguide/util/tasks/BaseRateItemTask.java
package com.battlelancer.seriesguide.util.tasks;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.SgApp;
import com.battlelancer.seriesguide.traktapi.SgTrakt;
import com.battlelancer.seriesguide.traktapi.TraktCredentials;
import com.battlelancer.seriesguide.util.Errors;
import com.uwetrottmann.trakt5.entities.SyncErrors;
import com.uwetrottmann.trakt5.entities.SyncItems;
import com.uwetrottmann.trakt5.entities.SyncResponse;
import com.uwetrottmann.trakt5.enums.Rating;
import com.uwetrottmann.trakt5.services.Sync;
import retrofit2.Response;

/**
 * Stores the rating in the database and sends it to Trakt.
 */
public abstract class BaseRateItemTask extends BaseActionTask {

    private final Rating rating;

    public BaseRateItemTask(Context context, Rating rating) {
        super(context);
        this.rating = rating;
    }

    @Override
    protected boolean isSendingToHexagon() {
        // Hexagon does not support ratings.
        return false;
    }

    @Override
    protected Integer doBackgroundAction(Void... params) {
        if (isSendingToTrakt()) {
            if (!TraktCredentials.get(getContext()).hasCredentials()) {
                return ERROR_TRAKT_AUTH;
            }

            SyncItems ratedItems = buildTraktSyncItems();
            if (ratedItems == null) {
                return ERROR_TRAKT_API;
            }
            SyncErrors notFound;
            try {
                Sync traktSync = SgApp.getServicesComponent(getContext()).traktSync();
                Response<SyncResponse> response = traktSync
                        .addRatings(ratedItems)
                        .execute();
                if (response.isSuccessful()) {
                    notFound = response.body().not_found;
                } else {
                    if (SgTrakt.isUnauthorized(getContext(), response)) {
                        return ERROR_TRAKT_AUTH;
                    }
                    Errors.logAndReport(getTraktAction(), response);
                    return ERROR_TRAKT_API;
                }
            } catch (Exception e) {
                Errors.logAndReport(getTraktAction(), e);
                return ERROR_TRAKT_API;
            }

            if (notFound != null) {
                if ((notFound.movies != null && notFound.movies.size() != 0)
                        || (notFound.shows != null && notFound.shows.size() != 0)
                        || (notFound.episodes != null && notFound.episodes.size() != 0)) {
                    // movie, show or episode not found on trakt
                    return ERROR_TRAKT_API_NOT_FOUND;
                }
            }
        }

        if (!doDatabaseUpdate()) {
            return ERROR_DATABASE;
        }

        return SUCCESS;
    }

    protected Rating getRating() {
        return rating;
    }

    @Override
    protected int getSuccessTextResId() {
        return R.string.trakt_success;
    }

    @NonNull
    protected abstract String getTraktAction();

    @Nullable
    protected abstract SyncItems buildTraktSyncItems();

    protected abstract boolean doDatabaseUpdate();
}

File: app/src/main/java/com/battlelancer/seriesguide/util/tasks/BaseRateItemTask.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2015-2024 Uwe Trottmann

package com.battlelancer.seriesguide.util.tasks

import android.content.Context
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.SgApp
import com.battlelancer.seriesguide.traktapi.TraktCredentials
import com.uwetrottmann.trakt5.entities.SyncItems
import com.uwetrottmann.trakt5.entities.SyncResponse
import com.uwetrottmann.trakt5.enums.Rating

/**
 * Stores the [rating] in the database and sends it to Trakt.
 *
 * If it is `null`, removes the rating instead.
 */
abstract class BaseRateItemTask(
    context: Context,
    protected val rating: Rating?
) : BaseActionTask(context) {

    override val isSendingToHexagon: Boolean
        get() = false // Hexagon does not support ratings.

    override fun doBackgroundAction(vararg params: Void?): Int? {
        if (isSendingToTrakt) {
            if (!TraktCredentials.get(context).hasCredentials()) {
                return ERROR_TRAKT_AUTH
            }

            val ratedItems = buildTraktSyncItems() ?: return ERROR_DATABASE

            val trakt = SgApp.getServicesComponent(context).trakt()
            val traktSync = trakt.sync()
            val call = if (rating != null) {
                traktSync.addRatings(ratedItems)
            } else {
                traktSync.deleteRatings(ratedItems)
            }
            val result =
                executeTraktCall<SyncResponse>(
                    call,
                    trakt,
                    traktAction,
                    object : ResponseCallback<SyncResponse> {
                        override fun handleSuccessfulResponse(body: SyncResponse): Int {
                            val notFound = body.not_found
                            if (notFound != null) {
                                val movies = notFound.movies
                                val shows = notFound.shows
                                val episodes = notFound.episodes
                                if ((movies != null && movies.size != 0)
                                    || (shows != null && shows.size != 0)
                                    || (episodes != null && episodes.size != 0)) {
                                    // movie, show or episode not found on trakt
                                    return ERROR_TRAKT_API_NOT_FOUND
                                }
                            }
                            return SUCCESS
                        }
                    })
            if (result != SUCCESS) {
                return result
            }
        }

        if (!doDatabaseUpdate()) {
            return ERROR_DATABASE
        }

        return SUCCESS
    }

    override val successTextResId: Int
        get() = R.string.ack_rated

    protected abstract val traktAction: String

    /**
     * May return null on database error.
     */
    protected abstract fun buildTraktSyncItems(): SyncItems?

    protected abstract fun doDatabaseUpdate(): Boolean
}
